### PR TITLE
[vidio] Add support for premier-exclusive videos

### PR DESCRIPTION
**Pull request by MinePlayersPE**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Add another API endpoint for downloading premier-only videos (requires account regardless of video, so old URL may still be used)

(also adds subtitle support)

### CODE REVIEW

File: yt_dlp/extractor/vidio.py [hunk 1]
```diff
@@ -56,9 +60,28 @@ def _real_extract(self, url):
...
+        hls_url = data['clips'][0]['hls_url']
+
+        formats, subs = self._extract_m3u8_formats_and_subtitles(
+            hls_url, display_id, 'mp4', 'm3u8_native', errnote=False, fatal=False)
+        if not formats:
+            self.to_screen('Falling back to premier mode.')
+            sources = self._download_json(
+                'https://www.vidio.com/interactions_stream.json?video_id=' + video_id + '&type=videos', display_id)
+            if not (sources.get('source') or sources.get('source_dash')):
+                self.raise_login_required()
```
Comment (pukkandan) [inline yt_dlp/extractor/vidio.py:72]:
```suggestion
                self.raise_login_required(method='cookies')
```

I had added an optional argument `method` to show a more usefull error message to user based on whether cookies or user-pass is the accepted auth method

File: yt_dlp/extractor/vidio.py [hunk 2]
```diff
@@ -56,9 +60,28 @@ def _real_extract(self, url):
...
+
+            if sources.get('source'):
+                hls_formats, hls_subs = self._extract_m3u8_formats_and_subtitles(
+                    sources['source'], display_id, 'mp4', 'm3u8_native')
+                formats.extend(hls_formats)
+                subs.update(hls_subs)
+            if sources.get('source_dash'):  # TODO: Find video example with source_dash, can both exist at the same time?
```
Comment (pukkandan) [inline yt_dlp/extractor/vidio.py:79]:
```suggestion
            if sources.get('source_dash'):  # TODO: Find video example with source_dash
```
> can both exist at the same time?

The code can already handle either case, so no need to put that in todo

File: yt_dlp/extractor/vidio.py [hunk 3]
```diff
@@ -56,9 +60,28 @@ def _real_extract(self, url):
             })
         video = data['videos'][0]
         title = video['title'].strip()
+        hls_url = data['clips'][0]['hls_url']
```
Comment (pukkandan) [inline yt_dlp/extractor/vidio.py:63]: is this key guaranteed to be present even for premium videos? If not, use `try_get`

**Review by pukkandan — APPROVED** [review]

**Comment by MinePlayersPE** [comment]
While looking at the JSON data, I noticed a key for checking premium videos. Revamped again for less hacky code

File: yt_dlp/extractor/vidio.py (hunk 1 above)
Comment (MinePlayersPE) [reply yt_dlp/extractor/vidio.py]: done

File: yt_dlp/extractor/vidio.py (hunk 2 above)
Comment (MinePlayersPE) [reply yt_dlp/extractor/vidio.py]: done

File: yt_dlp/extractor/vidio.py (hunk 3 above)
Comment (MinePlayersPE) [reply yt_dlp/extractor/vidio.py]: now irrelevant because premium is checked before `hls_url` is referenced

File: yt_dlp/extractor/vidio.py [hunk 4]
```diff
@@ -60,27 +60,29 @@ def _real_extract(self, url):
             })
         video = data['videos'][0]
         title = video['title'].strip()
-        hls_url = data['clips'][0]['hls_url']
-
-        formats, subs = self._extract_m3u8_formats_and_subtitles(
-            hls_url, display_id, 'mp4', 'm3u8_native', errnote=False, fatal=False)
-        if not formats:
-            self.to_screen('Falling back to premier mode.')
+        is_premium = video['is_premium']
```
Comment (pukkandan) [inline yt_dlp/extractor/vidio.py:63]:
```suggestion
        is_premium = video.get('is_premium')
```

File: yt_dlp/extractor/vidio.py [hunk 5]
```diff
@@ -56,9 +60,30 @@ def _real_extract(self, url):
             })
         video = data['videos'][0]
         title = video['title'].strip()
+        is_premium = video['is_premium']
+        if is_premium:
+            self.to_screen('Premier-exclusive video detected, switching to premier mode')
+            sources = self._download_json(
+                'https://www.vidio.com/interactions_stream.json?video_id=' + video_id + '&type=videos', display_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/vidio.py:67]:
```suggestion
            sources = self._download_json(
                'https://www.vidio.com/interactions_stream.json?video_id=%s&type=videos' % video_id,
                display_id, note='Downloading premium API JSON')
```

Feel free to change the actual `note` to whatever you want

File: yt_dlp/extractor/vidio.py [hunk 6]
```diff
@@ -56,9 +60,30 @@ def _real_extract(self, url):
             })
         video = data['videos'][0]
         title = video['title'].strip()
+        is_premium = video.get('is_premium')
+        if is_premium:
+            sources = sources = self._download_json(
```
Comment (pukkandan) [inline yt_dlp/extractor/vidio.py:65]:
```suggestion
            sources = self._download_json(
```

**Pull request merged**